--- a/ejabberd.yml.example
+++ b/ejabberd.yml.example
@@ -119,7 +119,7 @@
     ## file and uncomment these lines:
     ##
     certfile: 'CERTFILE'
-    starttls: true
+    ## starttls: true
     protocol_options: 'TLSOPTS'
     dhfile: 'DHFILE'
     ciphers: 'CIPHERS'
@@ -127,9 +127,9 @@
     ## To enforce TLS encryption for client connections,
     ## use this instead of the "starttls" option:
     ##
-    ## starttls_required: true
+    starttls_required: true
     ##
-    ## Stream compression
+    ## Custom OpenSSL options
     ##
     ## zlib: true
     ##
@@ -663,6 +663,7 @@
   mod_admin_extra: {}
   mod_announce: # recommends mod_adhoc
     access: announce
+  mod_block_strangers: {}
   mod_blocking: {} # requires mod_privacy
   mod_caps: {}
   mod_carboncopy: {}
@@ -671,7 +672,10 @@
   ## mod_delegation: {} # for xep0356
   mod_disco: {}
   mod_echo: {}
-  mod_irc: {}
+  mod_irc: 
+    default_encoding: "utf-8"
+  mod_http_bind:
+    max_inactivity: 60
   mod_bosh: {}
   ## mod_http_fileserver:
   ##   docroot: "/var/www"
@@ -702,9 +706,9 @@
   mod_offline: 
     access_max_user_messages: max_user_offline_messages
   mod_ping: {}
-  ## mod_pres_counter:
-  ##   count: 5
-  ##   interval: 60
+  mod_pres_counter:
+    count: 5
+    interval: 60
   mod_privacy: {}
   mod_private: {}
   ## mod_proxy65: {}
@@ -723,7 +727,7 @@
     ##
     ## Protect In-Band account registrations with CAPTCHA.
     ##
-    ##   captcha_protected: true
+    captcha_protected: true
     ##
     ## Set the minimum informational entropy for passwords.
     ##
@@ -754,6 +758,7 @@
     ##   access: register
   mod_roster: {}
   mod_shared_roster: {}
+  mod_sic: {}
   mod_stats: {}
   mod_time: {}
   mod_vcard:
