--- a/ejabberd.yml.example
+++ b/ejabberd.yml.example
@@ -59,7 +59,7 @@
 ##   - "example.org"
 ##
 hosts:
-  - "localhost"
+  - "__HOSTNAME__"
 
 ##
 ## route_subdomains: Delegate subdomains to other XMPP servers.
@@ -84,11 +84,13 @@
     ## certificate, specify the full path to the
     ## file and uncomment this line:
     ##
-    ## certfile: "/path/to/ssl.pem"
-    ## starttls: true
+    certfile: "/etc/ejabberd/ejabberd.pem"
+    starttls: true
+    starttls_required: true
     max_stanza_size: 65536
     shaper: c2s_shaper
     access: c2s
+    zlib: true
   - 
     port: 5269
     module: ejabberd_s2s_in
@@ -216,7 +218,7 @@
 ##
 ## List of LDAP servers:
 ## ldap_servers:
-##   - "localhost"
+##   - "ldap.__HOSTNAME__"
 ##
 ## Encryption of connection to LDAP servers:
 ## ldap_encrypt: none
@@ -346,8 +348,8 @@
   ##
   ## admin:
   ##   user:
-  ##     - "aleksey": "localhost"
-  ##     - "ermine": "example.org"
+  ##     - "aleksey": "__HOSTNAME__"
+  ##     - "ermine": "xmpp.example.com"
   ##
   ## Blocked users
   ##
@@ -393,11 +395,11 @@
 ## Define specific ACLs in a virtual host.
 ##
 ## host_config:
-##   "localhost":
+##   "__HOSTNAME__":
 ##     acl:
 ##       admin:
 ##         user:
-##           - "bob-local": "localhost"
+##           - "bob-local": "__HOSTNAME__"
 
 ###   ============
 ###   ACCESS RULES
@@ -461,7 +463,7 @@
 ## Define specific Access Rules in a virtual host.
 ##
 ## host_config:
-##   "localhost":
+##   "__HOSTNAME__":
 ##     access:
 ##       c2s:
 ##         admin: allow
@@ -481,7 +483,7 @@
 ## Set a different default language in a virtual host.
 ##
 ## host_config:
-##   "localhost":
+##   "__HOSTNAME__":
 ##     language: "ru"
 
 ###   =======
@@ -600,11 +602,11 @@
 ## Enable modules with custom options in a specific virtual host
 ##
 ## host_config:
-##   "localhost":
+##   "__HOSTNAME__":
 ##     add:
 ##       modules:
 ##         mod_echo:
-##           host: "mirror.localhost"
+##           host: "mirror.__HOSTNAME__"
 
 ### Local Variables:
 ### mode: yaml
