--- a/ejabberdctl.cfg.example
+++ b/ejabberdctl.cfg.example
@@ -135,3 +135,14 @@
 # Default: don't write PID file
 #
 #EJABBERD_PID_PATH=/var/run/ejabberd/ejabberd.pid
+
+# EJABBERD_MONIT_INTEGRATION: ejabberd integration into monit
+# 
+# Set to yes to enable integration into Monit[1]. Monit will watch ejabberd,
+# restart it if it crashes and sends you a mail on any status changes.
+#
+# [1] http://mmonit.com/monit/
+#
+# Default: no
+#
+#EJABBERD_MONIT_INTEGRATION=no
--- a/ejabberd.init.template
+++ b/ejabberd.init.template
@@ -22,6 +22,7 @@
 DIR=/usr/sbin
 CTL="$DIR"/ejabberdctl
 USER=@installuser@
+EJABBERD_MONIT_INTEGRATION=no
 
 # Include ejabberd defaults if available
 if [ -f /etc/default/ejabberd ] ; then
@@ -52,11 +53,13 @@
         su - $USER -c "$CTL start"
         su - $USER -c "$CTL started"
     fi
+    [ "$EJABBERD_MONIT_INTEGRATION" != 'yes' ] || monit monitor ejabberd
     echo "done."
     ;;
   stop)
     test -x "$CTL" || exit 0
     echo -n "Stopping ejabberd..."
+    [ "$EJABBERD_MONIT_INTEGRATION" != 'yes' ] || monit unmonitor ejabberd
     if status; then
         su - $USER -c "$CTL stop"
         su - $USER -c "$CTL stopped"
