--- a/ejabberdctl.template
+++ b/ejabberdctl.template
@@ -48,6 +48,7 @@
         --config-dir) ETCDIR=$1 ; shift ;;
         --config) EJABBERD_CONFIG_PATH=$1 ; shift ;;
         --ctl-config) EJABBERDCTL_CONFIG_PATH=$1 ; shift ;;
+        --default-dir) DEFAULTDIR=$1 ; shift ;;
         --logs) LOGS_DIR=$1 ; shift ;;
         --spool) SPOOLDIR=$1 ; shift ;;
         *) ARGS="$ARGS $PARAM" ;;
@@ -58,8 +59,11 @@
 if [ "$ETCDIR" = "" ] ; then
     ETCDIR={{sysconfdir}}/ejabberd
 fi
+if [ "$DEFAULTDIR" = "" ] ; then
+    DEFAULTDIR={{sysconfdir}}/default
+fi
 if [ "$EJABBERDCTL_CONFIG_PATH" = "" ] ; then
-    EJABBERDCTL_CONFIG_PATH=$ETCDIR/ejabberdctl.cfg
+    EJABBERDCTL_CONFIG_PATH=$DEFAULTDIR/ejabberdctl.cfg
 fi
 if [ -f "$EJABBERDCTL_CONFIG_PATH" ] ; then
     . "$EJABBERDCTL_CONFIG_PATH"
