--- a/ejabberd.init.template
+++ b/ejabberd.init.template
@@ -22,6 +22,7 @@
 CTL="$DIR"/ejabberdctl
 USER=ejabberd
 EJABBERDRUN=/var/run/ejabberd
+EJABBERD_PID_PATH=/run/ejabberd/ejabberd.pid
 
 test -x "$CTL" || {
 	echo "ERROR: ejabberd not found: $DIR"
@@ -43,12 +44,18 @@
 
 
 export PATH="${PATH:+$PATH:}/usr/sbin:/sbin"
+EJABBERD_PID_DIR=$(dirname $EJABBERD_PID_PATH)
 
 case "$1" in
   start)
     test -x "$CTL" || exit 0
     mkrundir
     echo "Starting ejabberd..."
+    if [ ! -d $EJABBERD_PID_DIR ]; then
+      mkdir $EJABBERD_PID_DIR
+      chown ejabberd:ejabberd $EJABBERD_PID_DIR
+      chmod 755 $EJABBERD_PID_DIR
+    fi
     if ! /etc/init.d/ejabberd status > /dev/null; then
       su - $USER -c "$CTL start"
       su - $USER -c "$CTL started"
@@ -62,6 +69,10 @@
       su - $USER -c "$CTL stop"
       su - $USER -c "$CTL stopped"
     fi
+    if [ -f $EJABBERD_PID_PATH ]; then
+        rm -f $EJABBERD_PID_PATH
+    fi
+
     echo "done."
     ;;
   status)
