--- a/ejabberd.service.template
+++ b/ejabberd.service.template
@@ -12,6 +12,7 @@
 RestartSec=5
 ExecStart=/bin/sh -c '/usr/sbin/ejabberdctl start && /usr/sbin/ejabberdctl started'
 ExecStop=/bin/sh -c '/usr/sbin/ejabberdctl stop && /usr/sbin/ejabberdctl stopped'
+ExecReload=/bin/sh -c '/usr/sbin/ejabberdctl reload_config'
 PIDFile=/run/ejabberd/ejabberd.pid
 PrivateDevices=true
 ProtectSystem=full
