--- a/ejabberd.yml.example
+++ b/ejabberd.yml.example
@@ -59,7 +59,7 @@
 ##   - "example.org"
 ##
 hosts:
-  - "localhost"
+  - "__HOSTNAME__"
 
 ##
 ## route_subdomains: Delegate subdomains to other XMPP servers.
@@ -84,11 +84,13 @@
     ## certificate, specify the full path to the
     ## file and uncomment this line:
     ##
-    ## certfile: "/path/to/ssl.pem"
-    ## starttls: true
+    certfile: "/etc/ejabberd/ejabberd.pem"
+    starttls: true
+    starttls_required: true
     max_stanza_size: 65536
     shaper: c2s_shaper
     access: c2s
+    zlib: true
   - 
     port: 5269
     module: ejabberd_s2s_in
@@ -137,12 +139,12 @@
 ## Allowed values are: false optional required required_trusted
 ## You must specify a certificate file.
 ##
-## s2s_use_starttls: optional
+s2s_use_starttls: required
 
 ##
 ## s2s_certfile: Specify a certificate file.
 ##
-## s2s_certfile: "/path/to/ssl.pem"
+s2s_certfile: "/etc/ejabberd/ejabberd.pem"
 
 ##
 ## domain_certfile: Specify a different certificate for each served hostname.
@@ -216,7 +218,7 @@
 ##
 ## List of LDAP servers:
 ## ldap_servers:
-##   - "localhost"
+##   - "ldap.__HOSTNAME__"
 ##
 ## Encryption of connection to LDAP servers:
 ## ldap_encrypt: none
@@ -346,8 +348,8 @@
   ##
   ## admin:
   ##   user:
-  ##     - "aleksey": "localhost"
-  ##     - "ermine": "example.org"
+  ##     - "aleksey": "__HOSTNAME__"
+  ##     - "ermine": "xmpp.example.com"
   ##
   ## Blocked users
   ##
@@ -393,11 +395,11 @@
 ## Define specific ACLs in a virtual host.
 ##
 ## host_config:
-##   "localhost":
+##   "__HOSTNAME__":
 ##     acl:
 ##       admin:
 ##         user:
-##           - "bob-local": "localhost"
+##           - "bob-local": "__HOSTNAME__"
 
 ###   ============
 ###   ACCESS RULES
@@ -453,6 +455,12 @@
   ##   bad_servers: deny
   ##   all: allow
 
+
+  max_s2s_connections:
+     all: 3
+  max_s2s_connections_per_node:
+     all: 3
+
 ## By default the frequency of account registrations from the same IP
 ## is limited to 1 account every 10 minutes. To disable, specify: infinity
 ## registration_timeout: 600
@@ -461,7 +469,7 @@
 ## Define specific Access Rules in a virtual host.
 ##
 ## host_config:
-##   "localhost":
+##   "__HOSTNAME__":
 ##     access:
 ##       c2s:
 ##         admin: allow
@@ -481,7 +489,7 @@
 ## Set a different default language in a virtual host.
 ##
 ## host_config:
-##   "localhost":
+##   "__HOSTNAME__":
 ##     language: "ru"
 
 ###   =======
@@ -490,12 +498,12 @@
 ##
 ## Full path to a script that generates the image.
 ##
-## captcha_cmd: "/lib/ejabberd/priv/bin/captcha.sh"
+captcha_cmd: "/usr/lib/ejabberd/priv/bin/captcha.sh"
 
 ##
 ## Host for the URL and port where ejabberd listens for CAPTCHA requests.
 ##
-## captcha_host: "example.org:5280"
+captcha_host: "https://__HOSTNAME__:5280"
 
 ##
 ## Limit CAPTCHA calls per minute for JID/IP to avoid DoS.
@@ -510,15 +518,22 @@
 ##
 modules: 
   mod_adhoc: {}
+  mod_admin_extra: {} # enable extra admin commands
   mod_announce: # recommends mod_adhoc
     access: announce
   mod_blocking: {} # requires mod_privacy
   mod_caps: {}
   mod_configure: {} # requires mod_adhoc
-  mod_disco: {}
+  mod_disco:
+    server_info:
+        modules: all
+        name: "abuse-addresses"
+        urls: ["mailto:abuse@__HOSTNAME__"]
   ## mod_echo: {}
-  mod_irc: {}
-  mod_http_bind: {}
+  mod_irc:
+    default_encoding: "utf-8"
+  mod_http_bind:
+    max_inactivity: 60
   ## mod_http_fileserver:
   ##   docroot: "/var/www"
   ##   accesslog: "/var/log/ejabberd/access.log"
@@ -529,13 +544,15 @@
     access_create: muc_create
     access_persistent: muc_create
     access_admin: muc_admin
+    min_message_interval: 1.0
+    min_presence_interval: 5.0
   ## mod_muc_log: {}
   mod_offline: 
     access_max_user_messages: max_user_offline_messages
   mod_ping: {}
-  ## mod_pres_counter:
-  ##   count: 5
-  ##   interval: 60
+  mod_pres_counter:
+     count: 5
+     interval: 60
   mod_privacy: {}
   mod_private: {}
   ## mod_proxy65: {}
@@ -554,7 +571,7 @@
     ##
     ## Protect In-Band account registrations with CAPTCHA.
     ##
-    ## captcha_protected: true
+    captcha_protected: true
 
     ##
     ## Set the minimum informational entropy for passwords.
@@ -581,7 +598,7 @@
     ##
     ## Only clients in the server machine can register accounts
     ##
-    ip_access: trusted_network
+    ## ip_access: trusted_network
 
     ##
     ## Local c2s or remote s2s users cannot register accounts
@@ -590,6 +607,7 @@
 
     access: register
   mod_roster: {}
+  mod_sic: {}
   mod_shared_roster: {}
   mod_stats: {}
   mod_time: {}
@@ -600,11 +618,11 @@
 ## Enable modules with custom options in a specific virtual host
 ##
 ## host_config:
-##   "localhost":
+##   "__HOSTNAME__":
 ##     add:
 ##       modules:
 ##         mod_echo:
-##           host: "mirror.localhost"
+##           host: "mirror.__HOSTNAME__"
 
 ### Local Variables:
 ### mode: yaml
--- a/ejabberdctl.cfg.example
+++ b/ejabberdctl.cfg.example
@@ -3,7 +3,7 @@
 # to the erlang runtime system when starting ejabberd
 #
 
-#' POLL: Kernel polling ([true|false])
+# POLL: Kernel polling ([true|false])
 #
 # The kernel polling option requires support in the kernel.
 # Additionally, you need to enable this feature while compiling Erlang.
@@ -12,8 +12,7 @@
 #
 #POLL=true
 
-#.
-#' SMP: SMP support ([enable|auto|disable])
+# SMP: SMP support ([enable|auto|disable])
 #
 # Explanation in Erlang/OTP documentation:
 # enable: starts the Erlang runtime system with SMP support enabled.
@@ -26,8 +25,7 @@
 #
 #SMP=auto
 
-#.
-#' ERL_MAX_PORTS: Maximum number of simultaneously open Erlang ports
+# ERL_MAX_PORTS: Maximum number of simultaneously open Erlang ports
 #
 # ejabberd consumes two or three ports for every connection, either
 # from a client or from another Jabber server. So take this into
@@ -38,8 +36,7 @@
 #
 #ERL_MAX_PORTS=32000
 
-#.
-#' FIREWALL_WINDOW: Range of allowed ports to pass through a firewall
+# FIREWALL_WINDOW: Range of allowed ports to pass through a firewall
 #
 # If Ejabberd is configured to run in cluster, and a firewall is blocking ports,
 # it's possible to make Erlang use a defined range of port (instead of dynamic
@@ -50,8 +47,7 @@
 #
 #FIREWALL_WINDOW=
 
-#.
-#' INET_DIST_INTERFACE: IP address where this Erlang node listens other nodes
+# INET_DIST_INTERFACE: IP address where this Erlang node listens other nodes
 #
 # This communication is used by ejabberdctl command line tool,
 # and in a cluster of several ejabberd nodes.
@@ -60,10 +56,7 @@
 #
 #INET_DIST_INTERFACE=127.0.0.1
 
-#.
-#' ERL_EPMD_ADDRESS: IP addresses where epmd listens for connections
-#
-# IMPORTANT: This option works only in Erlang/OTP R14B03 and newer.
+# ERL_EPMD_ADDRESS: IP addresses where epmd listens for connections
 #
 # This environment variable may be set to a comma-separated
 # list of IP addresses, in which case the epmd daemon
@@ -76,8 +69,7 @@
 #
 #ERL_EPMD_ADDRESS=127.0.0.1
 
-#.
-#' ERL_PROCESSES: Maximum number of Erlang processes
+# ERL_PROCESSES: Maximum number of Erlang processes
 #
 # Erlang consumes a lot of lightweight processes. If there is a lot of activity
 # on ejabberd so that the maximum number of processes is reached, people will
@@ -90,8 +82,7 @@
 #
 #ERL_PROCESSES=250000
 
-#.
-#' ERL_MAX_ETS_TABLES: Maximum number of ETS and Mnesia tables
+# ERL_MAX_ETS_TABLES: Maximum number of ETS and Mnesia tables
 #
 # The number of concurrent ETS and Mnesia tables is limited. When the limit is
 # reached, errors will appear in the logs:
@@ -103,8 +94,7 @@
 #
 #ERL_MAX_ETS_TABLES=1400
 
-#.
-#' ERL_OPTIONS: Additional Erlang options
+# ERL_OPTIONS: Additional Erlang options
 #
 # The next variable allows to specify additional options passed to erlang while
 # starting ejabberd. Some useful options are -noshell, -detached, -heart. When
@@ -118,8 +108,7 @@
 #
 #ERL_OPTIONS=""
 
-#.
-#' ERLANG_NODE: Erlang node name
+# ERLANG_NODE: Erlang node name
 #
 # The next variable allows to explicitly specify erlang node for ejabberd
 # It can be given in different formats:
@@ -136,8 +125,7 @@
 #
 #ERLANG_NODE=ejabberd@localhost
 
-#.
-#' EJABBERD_PID_PATH: ejabberd PID file
+# EJABBERD_PID_PATH: ejabberd PID file
 #
 # Indicate the full path to the ejabberd Process identifier (PID) file.
 # If this variable is defined, ejabberd writes the PID file when starts,
@@ -147,7 +135,3 @@
 # Default: don't write PID file
 #
 #EJABBERD_PID_PATH=/var/run/ejabberd/ejabberd.pid
-
-#.
-#'
-# vim: foldmarker=#',#. foldmethod=marker:
