Author: Gerfried Fuchs <rhonda@debian.at>	vim:ft=diff:
Description: Fix syntax problems with the example scripts

--- a/examples/mtr/ejabberd
+++ b/examples/mtr/ejabberd
@@ -44,9 +44,8 @@
         ;;
 esac
 
-if checkyesno ${rcvar}
+if ! checkyesno ${rcvar}
 then
-else
     exit 0
 fi
 
--- a/examples/mtr/ejabberd-netbsd.sh
+++ b/examples/mtr/ejabberd-netbsd.sh
@@ -8,8 +8,7 @@
     make
     make install
 fi
-if pkg_info erlang | grep -q erlang-9.1nb1; then
-else
+if ! pkg_info erlang | grep -q erlang-9.1nb1; then
     echo "erlang-9.1nb1 not installed" 1>&2
     exit 1
 fi
--- a/ejabberdctl.template
+++ b/ejabberdctl.template
@@ -7,14 +7,14 @@
 ERL_PROCESSES=250000
 ERL_MAX_ETS_TABLES=1400
 FIREWALL_WINDOW=""
-ERLANG_NODE=ejabberd@localhost
+ERLANG_NODE=ejabberd
 
 # define default environment variables
 SCRIPT_DIR=`cd ${0%/*} && pwd`
 ERL={{erl}}
 IEX={{bindir}}/iex
 EPMD={{bindir}}/epmd
-INSTALLUSER={{installuser}}
+INSTALLUSER=ejabberd
 
 # Compatibility in ZSH
 #setopt shwordsplit 2>/dev/null
