--- a/src/ext_mod.erl
+++ b/src/ext_mod.erl
@@ -468,6 +468,7 @@
     Logger = [{d, 'LAGER'} || code:is_loaded(lager)=/=false],
     ExtLib = [{d, 'NO_EXT_LIB'} || filelib:is_file(XmlHrl)],
     Options = [{outdir, Ebin}, {i, "include"}, {i, EjabInc},
+               {d, 'NO_EXT_LIB'},  %% use include instead of include_lib
                verbose, report_errors, report_warnings]
               ++ Logger ++ ExtLib,
     Result = [case compile:file(File, Options) of
