--- a/src/ext_mod.erl
+++ b/src/ext_mod.erl
@@ -446,7 +446,6 @@
         _ -> []
     end,
     Options = [{outdir, Ebin}, {i, "include"}, {i, EjabInc},
-               {d, 'NO_EXT_LIB'},  %% use include instead of include_lib
                verbose, report_errors, report_warnings]
               ++ Logger,
     Result = [case compile:file(File, Options) of
--- a/ejabberd.service.template
+++ b/ejabberd.service.template
@@ -7,9 +7,9 @@
 Group=ejabberd
 LimitNOFILE=16000
 RestartSec=5
-ExecStart=/bin/sh @ctlscriptpath@/ejabberdctl start
-ExecStop=@ctlscriptpath@/ejabberdctl stop
-ExecReload=@ctlscriptpath@/ejabberdctl reload_config
+ExecStart=/bin/sh /usr/sbin/ejabberdctl start
+ExecStop=/usr/sbin/ejabberdctl stop
+ExecReload=/usr/sbin/ejabberdctl reload_config
 Type=oneshot
 RemainAfterExit=yes
 
