From 6c712b6caaaed35471452612d643f785bfdf11ee Mon Sep 17 00:00:00 2001
From: Holger Weiss <holger@zedat.fu-berlin.de>
Date: Thu, 4 Jun 2015 09:55:38 +0200
Subject: [PATCH] Don't add body element to MUC subject messages

---
 src/mod_muc_room.erl | 30 ++++++++++--------------------
 1 file changed, 10 insertions(+), 20 deletions(-)

diff --git a/src/mod_muc_room.erl b/src/mod_muc_room.erl
index c52b4e3..f28fe00 100644
--- a/src/mod_muc_room.erl
+++ b/src/mod_muc_room.erl
@@ -1883,7 +1883,7 @@ add_new_user(From, Nick,
 		Shift = count_stanza_shift(Nick, Els, NewState),
 		case send_history(From, Shift, NewState) of
 		  true -> ok;
-		  _ -> send_subject(From, Lang, StateData)
+		  _ -> send_subject(From, StateData)
 		end,
 		case NewState#state.just_created of
 		  true -> NewState#state{just_created = false};
@@ -2502,25 +2502,15 @@ send_history(JID, Shift, StateData) ->
 		lists:nthtail(Shift,
 			      lqueue_to_list(StateData#state.history))).
 
-send_subject(JID, Lang, StateData) ->
-    case StateData#state.subject_author of
-      <<"">> -> ok;
-      Nick ->
-	  Subject = StateData#state.subject,
-	  Packet = #xmlel{name = <<"message">>,
-			  attrs = [{<<"type">>, <<"groupchat">>}],
-			  children =
-			      [#xmlel{name = <<"subject">>, attrs = [],
-				      children = [{xmlcdata, Subject}]},
-			       #xmlel{name = <<"body">>, attrs = [],
-				      children =
-					  [{xmlcdata,
-					    <<Nick/binary,
-					      (translate:translate(Lang,
-								   <<" has set the subject to: ">>))/binary,
-					      Subject/binary>>}]}]},
-	  ejabberd_router:route(StateData#state.jid, JID, Packet)
-    end.
+send_subject(_JID, #state{subject_author = <<"">>}) -> ok;
+send_subject(JID, StateData) ->
+    Subject = StateData#state.subject,
+    Packet = #xmlel{name = <<"message">>,
+		    attrs = [{<<"type">>, <<"groupchat">>}],
+		    children =
+			[#xmlel{name = <<"subject">>, attrs = [],
+				children = [{xmlcdata, Subject}]}]},
+    ejabberd_router:route(StateData#state.jid, JID, Packet).
 
 check_subject(Packet) ->
     case xml:get_subtag(Packet, <<"subject">>) of
