--- a/ejabberdctl.cfg.example
+++ b/ejabberdctl.cfg.example
@@ -121,9 +121,9 @@
 #   The same as previous, but erlang will use long hostname
 #   (see erl (1) manual for details)
 #
-# Default: ejabberd@localhost
+# Default: ejabberd
 #
-#ERLANG_NODE=ejabberd@localhost
+#ERLANG_NODE=ejabberd
 
 # EJABBERD_PID_PATH: ejabberd PID file
 #
--- a/ejabberdctl.template
+++ b/ejabberdctl.template
@@ -7,7 +7,7 @@
 ERL_PROCESSES=250000
 ERL_MAX_ETS_TABLES=1400
 FIREWALL_WINDOW=""
-ERLANG_NODE=ejabberd@localhost
+ERLANG_NODE=ejabberd
 EJABBERD_PID_PATH=/run/ejabberd/ejabberd.pid
 
 # define default environment variables
@@ -293,7 +293,7 @@
         JOT='/usr/bin/jot'
         if [ ! -x "$JOT" ] ; then
             # no flock or jot, simply invoke ctlexec()
-            CTL_CONN="ctl-${ERLANG_NODE}"
+            CTL_CONN="ctl-$$-${ERLANG_NODE}"
             ctlexec $CTL_CONN $COMMAND
             result=$?
         else
