From fda73c3d163cb0bf26c011c62affdce8a2f5098e Mon Sep 17 00:00:00 2001
From: Evgeniy Khramtsov <ekhramtsov@process-one.net>
Date: Mon, 1 Sep 2014 16:38:41 +0400
Subject: [PATCH] Fix privacy checks of presence probes

---
 src/ejabberd_c2s.erl | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

--- a/src/ejabberd_c2s.erl
+++ b/src/ejabberd_c2s.erl
@@ -2194,7 +2194,7 @@
 	   [],
 	   StateData#state.pres_t),
     PacketProbe = #xmlel{name = <<"presence">>, attrs = [{<<"type">>,<<"probe">>}], children = []},
-    JIDs2Probe = format_and_check_privacy(From, StateData, Packet, JIDsProbe, out),
+    JIDs2Probe = format_and_check_privacy(From, StateData, PacketProbe, JIDsProbe, out),
     Server = StateData#state.server,
     send_multiple(StateData, From, JIDs2Probe, PacketProbe),
     {As, JIDs} =
