--- a/rebar.config.script
+++ b/rebar.config.script
@@ -53,13 +53,13 @@
                     Acc
             end, [], Cfg),
 
-Deps = [{p1_cache_tab, ".*", {git, "git://github.com/processone/cache_tab"}},
-        {p1_tls, ".*", {git, "git://github.com/processone/tls"}},
-        {p1_stringprep, ".*", {git, "git://github.com/processone/stringprep"}},
-        {p1_xml, ".*", {git, "git://github.com/processone/xml"}},
-	{esip, ".*", {git, "git://github.com/processone/p1_sip"}},
-        {p1_yaml, ".*", {git, "git://github.com/processone/p1_yaml"}},
-        {xmlrpc, ".*", {git, "git://github.com/rds13/xmlrpc"}}],
+Deps = [{p1_cache_tab, ".*", {git, "git://github.com/processone/cache_tab", "d8033a305d4b47ca0b7ba770a1bab5bae066c748"}},
+        {p1_tls, ".*", {git, "git://github.com/processone/tls", "f6aef446b5d4a0f8d18d2afefd043d2e3d87093b"}},
+        {p1_stringprep, ".*", {git, "git://github.com/processone/stringprep", "9e9e0f8dbe6a70ef36e1d4436b458ca5a77fbcfb"}},
+        {p1_xml, ".*", {git, "git://github.com/processone/xml", "73fa96fec08a20c2b57cc0105644b3c63aa56533"}},
+	{esip, ".*", {git, "git://github.com/processone/p1_sip", "ca05b2dc11fb5499cd3be080692f3239d7fb5c73"}},
+        {p1_yaml, ".*", {git, "git://github.com/processone/p1_yaml", "da07a8aa54a830acf319747a9ddabe3ac74f8490"}},
+        {xmlrpc, ".*", {git, "git://github.com/rds13/xmlrpc", "9e9be055c248be7bfb87b04dd9c444bc46e54912"}}],
 
 ConfigureCmd = fun(Pkg, Flags) ->
                        {'get-deps',
@@ -84,26 +84,26 @@
 
 CfgDeps = lists:flatmap(
             fun({mysql, true}) ->
-                    [{p1_mysql, ".*", {git, "git://github.com/processone/mysql"}}];
+                    [{p1_mysql, ".*", {git, "git://github.com/processone/mysql", "52e28c8df7dcf1d3e9de6ac9def65f79e2f864dc"}}];
                ({pgsql, true}) ->
-                    [{p1_pgsql, ".*", {git, "git://github.com/processone/pgsql"}}];
+                    [{p1_pgsql, ".*", {git, "git://github.com/processone/pgsql", "58a09c5cb94a59ffb7a79049e16d2ed9701f3195"}}];
                ({pam, true}) ->
-                    [{p1_pam, ".*", {git, "git://github.com/processone/epam"}}];
+                    [{p1_pam, ".*", {git, "git://github.com/processone/epam", "99c6705469fcb5de5186d33a15e8c15134a91727"}}];
                ({zlib, true}) ->
-                    [{p1_zlib, ".*", {git, "git://github.com/processone/zlib"}}];
+                    [{p1_zlib, ".*", {git, "git://github.com/processone/zlib", "52e82bbc12258ca17ea7ed9b5d10eca60c3db006"}}];
                ({stun, true}) ->
-                    [{p1_stun, ".*", {git, "git://github.com/processone/stun"}}];
+                    [{p1_stun, ".*", {git, "git://github.com/processone/stun", "1639c51d2c315ec03b541f4079b15565ed0b61f6"}}];
                ({json, true}) ->
-                    [{jiffy, ".*", {git, "git://github.com/davisp/jiffy"}}];
+                    [{jiffy, ".*", {git, "git://github.com/davisp/jiffy", "1e95a84b0ce8d5a86e3431e2095f245272c88662"}}];
                ({iconv, true}) ->
-                    [{p1_iconv, ".*", {git, "git://github.com/processone/eiconv"}}];
+                    [{p1_iconv, ".*", {git, "git://github.com/processone/eiconv", "7afc233d484566d8f87e0cbedd514a472626caf8"}}];
                ({http, true}) ->
-                    [{ibrowse, ".*", {git, "git://github.com/cmullaparthi/ibrowse"}},
-                     {lhttpc, ".*", {git, "git://github.com/esl/lhttpc"}}];
+                    [{ibrowse, ".*", {git, "git://github.com/cmullaparthi/ibrowse", "5bae7308a749f2dc801347c27f56dc1a21996aea"}},
+                     {lhttpc, ".*", {git, "git://github.com/esl/lhttpc", "5e459ec93f916ca7e5f2e1d4b14a20fbb0c49402"}}];
                ({lager, true}) ->
-                    [{lager, ".*", {git, "git://github.com/basho/lager"}}];
+                    [{lager, ".*", {git, "git://github.com/basho/lager", "4d8291edb72a25ac0efde907e2ecaa115108ff90"}}];
                ({lager, false}) ->
-                    [{p1_logger, ".*", {git, "git://github.com/processone/p1_logger"}}];
+                    [{p1_logger, ".*", {git, "git://github.com/processone/p1_logger", "c7141aff68282dec46d60a3f39cbf658d0c91961"}}];
                (_) ->
                     []
             end, Cfg),
