--- a/ejabberd.service.template
+++ b/ejabberd.service.template
@@ -5,13 +5,13 @@
 [Service]
 User=ejabberd
 Group=ejabberd
-LimitNOFILE=16000
+LimitNOFILE=32768
+Type=forking
+Restart=on-failure
 RestartSec=5
-ExecStart=/bin/sh @ctlscriptpath@/ejabberdctl start
-ExecStop=@ctlscriptpath@/ejabberdctl stop
-ExecReload=@ctlscriptpath@/ejabberdctl reload_config
-Type=oneshot
-RemainAfterExit=yes
+ExecStart=/usr/sbin/ejabberdctl start
+ExecStop=/usr/sbin/ejabberdctl stop
+ExecReload=/usr/sbin/ejabberdctl reload_config
 
 [Install]
 WantedBy=multi-user.target
